server/service/src/dashboard/item_count.rs

```diff
@@ -0,0 +1,278 @@
+use std::convert::TryInto;
+
+use repository::{ItemFilter, ItemRepository, RepositoryError};
+
+use crate::{item_stats::get_item_stats, service_provider::ServiceContext};
+
+pub struct ItemCounts {
+    pub total: i64,
+    pub no_stock: i64,
+    pub low_stock: i64,
+}
+
+pub trait ItemCountServiceTrait: Send + Sync {
+    /// # Arguments
+    ///
+    /// * i32 threshold number of months below which is considered low stock
+    fn get_item_counts(
+        &self,
+        ctx: &ServiceContext,
+        store_id: &str,
+        low_stock_threshold: i32,
+    ) -> Result<ItemCounts, RepositoryError> {
+        ItemServiceCount {}.get_item_counts(ctx, store_id, low_stock_threshold)
+    }
+}
+
+pub struct ItemServiceCount {}
+
+impl ItemCountServiceTrait for ItemServiceCount {
+    fn get_item_counts(
+        &self,
+        ctx: &ServiceContext,
+        store_id: &str,
+        low_stock_threshold: i32,
+    ) -> Result<ItemCounts, RepositoryError> {
+        let repo = ItemRepository::new(&ctx.connection);
+        let total = match repo.count(Some(ItemFilter::new().match_is_visible(true))) {
+            Ok(total) => total,
+            Err(error) => return Err(error),
+        };
+
+        let no_stock =
+            match repo.count_no_stock(Some(ItemFilter::new().match_is_visible(true)), store_id) {
+                Ok(no_stock) => no_stock,
+                Err(error) => return Err(error),
+            };
+
+        let item_stats = get_item_stats(ctx, store_id, None, None)
+            .unwrap()
+            .into_iter()
+            .filter(|item| {
+                item.average_monthly_consumption != 0.0
+                    && item.available_stock_on_hand as f64 / item.average_monthly_consumption
+                        < low_stock_threshold as f64
+            });
+
+        let low_stock: i64 = match item_stats.count().try_into() {
+            Ok(low_stock) => low_stock,
+            Err(error) => {
+                return Err(RepositoryError::DBError {
+                    msg: error.to_string(),
+                    extra: "".to_string(),
+                })
+            }
+        };
+
+        Ok(ItemCounts {
+            total,
+            no_stock,
+            low_stock,
+        })
+    }
+}
+
+#[cfg(test)]
+mod item_count_service_test {
+    use chrono::{Duration, Utc};
+    use repository::{
+        mock::{mock_master_list_item_query_test1, mock_store_b, MockDataInserts},
+        test_db, InvoiceLineRow, InvoiceLineRowRepository, InvoiceLineRowType, InvoiceRow,
+        InvoiceRowRepository, InvoiceRowType, ItemRow, ItemRowRepository, ItemRowType,
+        MasterListLineRow, MasterListLineRowRepository, StockLineRow, StockLineRowRepository,
+    };
+    use util::inline_init;
+
+    use crate::{
+        dashboard::item_count::{ItemCountServiceTrait, ItemServiceCount},
+        service_provider::ServiceContext,
+    };
+
+    fn data(
```

Comment:
nit: Ideally test data should be quite static and easy to digest (if it's possible to do add data like that statically). 
Personally in this case I wouldn't worry about large set of data, just enough to test the use cases (like 5), and static data:

```rust
let ServiceTestContext {
            service_provider,
            processors_task,
            ..
        } = setup_all_with_data_and_service_provider(
            "invoice_transfers",
            MockDataInserts::none().stores().names(),
            MockData {
                items: vec![
                    ItemRow {
                        id: "item1".to_string(),
                        r#type: ItemRowType::Stock,
                        ..ItemRow::default()
                    },
                    ItemRow {
                        id: "item2".to_string(),
                        r#type: ItemRowType::Stock,
                        ..ItemRow::default()
                    },
                    ItemRow {
                        id: "item3".to_string(),
                        r#type: ItemRowType::Stock,
                        ..ItemRow::default()
                    },
                    ItemRow {
                        id: "item4".to_string(),
                        r#type: ItemRowType::Stock,
                        ..ItemRow::default()
                    },
                    ItemRow {
                        id: "item5".to_string(),
                        r#type: ItemRowType::Stock,
                        ..ItemRow::default()
                    },
                ],
                full_master_lists: vec![FullMockMasterList {
                    master_list: MasterListRow {
                        id: "list1".to_string(),
                        ..MasterListRow::default()
                    },
                    joins: vec![MasterListNameJoinRow {
                        id: "join1".to_string(),
                        master_list_id: "list1".to_string(),
                        name_id: mock_store_b().name_id,
                    }],
                    lines: vec![
                        MasterListLineRow {
                            id: "listline1".to_string(),
                            item_id: "item1".to_string(),
                            master_list_id: "list1".to_string(),
                        },
                        MasterListLineRow {
                            id: "listline2".to_string(),
                            item_id: "item2".to_string(),
                            master_list_id: "list1".to_string(),
                        },
                        MasterListLineRow {
                            id: "listline3".to_string(),
                            item_id: "item3".to_string(),
                            master_list_id: "list1".to_string(),
                        },
                        MasterListLineRow {
                            id: "listline4".to_string(),
                            item_id: "item4".to_string(),
                            master_list_id: "list1".to_string(),
                        },
                    ],
                }],
                stock_lines: vec![
                    StockLineRow {
                        id: "stock_line1".to_string(),
                        item_id: "item1".to_string(),
                        store_id: mock_store_b().id,
                        ..StockLineRow::default()
                    },
                    // Can easiy add multiple stock lines and don't have to remove stock lines can just not add them
                    // Etc + invoice and invoice lines
                ],
                ..MockData::default()
            },
        )
        .await;
```

Also notice `setup_all_with_data_and_service_provider` and `setup_all_and_service_provider` can be used to get service provider and service context.

And although this seems like it's quite a bit of typing, it's actually quite easy, copy and paste once first record is added.

Reply:
Also using static way of creating mock data allows easier way to create variations of test data, in this case I think count_no_stock implementation needs adjustment (it doesn't count stock_on_hand for item, but number of stock_lines that have 0 available quantity).

This test fails: https://github.com/openmsupply/open-msupply/compare/927-item-stats...927-example-extra-test